cmd/activator/main.go

```diff
@@ -162,29 +161,21 @@ func main() {
 	// Enable TLS against queue-proxy when internal-encryption is enabled.
 	tlsEnabled := networkConfig.InternalEncryption
 
+	var certCache *certificate.CertCache
+	if tlsEnabled {
+		logger.Info("Internal Encryption is enabled")
+		certCache = certificate.NewCertCache(ctx)
+	}
+
 	// Enable TLS client when queue-proxy-ca is specified.
 	// At this moment activator with TLS does not disable HTTP.
 	// See also https://github.com/knative/serving/issues/12808.
 	if tlsEnabled {
-		logger.Info("Internal Encryption is enabled")
-		caSecret, err := kubeClient.CoreV1().Secrets(system.Namespace()).Get(ctx, netcfg.ServingInternalCertName, metav1.GetOptions{})
-		if err != nil {
-			logger.Fatalw("Failed to get secret", zap.Error(err))
-		}
-
-		pool, err := x509.SystemCertPool()
-		if err != nil {
-			pool = x509.NewCertPool()
-		}
-
-		if ok := pool.AppendCertsFromPEM(caSecret.Data[certificates.CaCertName]); !ok {
-			logger.Fatalw("Failed to append ca cert to the RootCAs")
-		}
-
 		tlsConf := &tls.Config{
-			RootCAs:            pool,
-			InsecureSkipVerify: false,
-			ServerName:         certificates.FakeDnsName,
+			VerifyPeerCertificate: certCache.ValidateCert,
+			// nolint:gosec
+			// Validate certificates by a custom function via VerifyPeerCertificate.
+			InsecureSkipVerify: true,
```

Comment:
1. Best to avoid replacing the golang tls lib verification with our own unless we are compelled to do that. Better to try to use the library tls verification (and any future updates it has) and add verifications as needed.
2. Can we find a way to update tls.Config with the changes on every connection attempt made (or find a way to update the used tls.Config on any change we detect?)   
3. Instead of (3), can we change our install to ensure certificates are created before we get started and have the certificates created regardless if encryption is used, then require that the activator is restarted when a CA changes?

Reply:
For 2)
As far as I searched, it is not possible. e.g. https://github.com/golang/go/issues/41888 https://github.com/golang/go/issues/35887
And actually using `VerifyPeerCertificate` was recommended by a Go developer - https://github.com/golang/go/issues/22836?utm_source=pocket_saves#issuecomment-584968697

For 3)
It is possible to restart activator (send TERM signal) when the CA is updated. But is it better to restart? As said 2), `VerifyPeerCertificate` was recommended.

Reply:
Another approach is to find a way to catch when the client fails to connect and based on the failure see if we need to update the CA.

Reply:
> Another approach is to find a way to catch when the client fails to connect and based on the failure see if we need to update the CA.

It is an approach for "how to detect the cert expiration" not "how to re-load the updated certs", right? Current approach (checking the secrets by informer) is better that that, I think.

Reply:
@nak3 

> It is an approach for "how to detect the cert expiration" not "how to re-load the updated certs", right? Current approach (checking the secrets by informer) is better that that, I think.

Agree

As for "how to re-load the updated certs", note that tlsConf is a pointer to the tls.Config struct. 
If you add one to CertCache struct, you can use a pointer to it instead of the one created in main.go, you can then apparently update it whenever you want. 

cache.go:
```golang 
// CertCache caches certificates and tls.Config.
type CertCache struct {
	secretInformer v1.SecretInformer
	logger         *zap.SugaredLogger

	certificates *tls.Certificate
	tlsConf         tls.Config

	certificatesMux sync.RWMutex
}

...
func (cr *CertCache) handleCertificateAdd(added interface{}) {
	if secret, ok := added.(*corev1.Secret); ok {
		...
		pool.AddCert(ca)

		cr.tlsConf.RootsCA = pool
	}
}

```

main.go:
```golang
transport = pkgnet.NewProxyAutoTLSTransport(env.MaxIdleProxyConns, env.MaxIdleProxyConnsPerHost, &certCache.tlsConf)
```

I did not test this with your code, and there are some missing pieces like initializing the tlcConf with the regular:
```golang
     InsecureSkipVerify: false,
     ServerName:         certificates.FakeDnsName,
     MinVersion: tls.VersionTLS12,
```

While testing this, I noticed that we fail to reconcile namespace secrets properly, I had to delete the secret manually and recreate the service every time... we also need to look into that...

Reply:
Assuming proper locking, replacing the set of root CA certs seems feasible.  I'd want to test the behavior of replacing the cert *while* a handshake is ongoing; alternatively, we could cache the entire client and replace it, allowing go's GC to collect the old client once all usages have completed.

Reply:
Do we really need locking here?
RootCAs is a set of certificates. Either the certificate from the server is included in this set or not.

- If the certificate is included in this set just before the replacement but not after - it is ok if we open the TLS connection and it is ok if we don't.
- If the certificate is included in this set just after the replacement but not before - it is ok if we open the TLS connection and it is ok if we don't.
- If the certificate is included in this set before and after - it is ok to open the TLS connection apparently we will... 
- If the certificate is _not_ included in this set either before or after - we must block the connection and apparently we will...

Reply:
@davidhadas Does your suggestion https://github.com/knative/serving/pull/13854#discussion_r1161068967 work on your env? Although it seems you are testing on your env, I think it does not work. I demonstrated the change on this PR and CI fails - see https://github.com/knative/serving/actions/runs/4653632630/jobs/8234731917?pr=13854 so could you point out what is different from your testing?

Reply:
Thank you.
Okay, so we should pass call `NewProxyAutoTLSTransport`

```go
transport = pkgnet.NewProxyAutoTLSTransport(env.MaxIdleProxyConns, env.MaxIdleProxyConnsPerHost, &certCache.tlsConf)
```
instead of:

```go
transport = pkgnet.NewProxyAutoTLSTransport(env.MaxIdleProxyConns, env.MaxIdleProxyConnsPerHost, certCache.tlsConf)
```

as you are suggesting `certCache.GetTLSConfig()` should return pointer?

Reply:
Thanks to the change you made to GetTLSConfig it now returns a pointer. 

So tlsConf is now of type *tls.Config:
```
tlsConf := certCache.GetTLSConfig()
```

and therefore it is fine to have:
```
     transport = pkgnet.NewProxyAutoTLSTransport(env.MaxIdleProxyConns, env.MaxIdleProxyConnsPerHost, tlsConf)
```

alternatively, you can get rid of `tlsConf` and of `GetTLSConfig` and just use:
```
    transport = pkgnet.NewProxyAutoTLSTransport(env.MaxIdleProxyConns, env.MaxIdleProxyConnsPerHost, &certCache.tlsConf)
```

Slack me if this is still confusing.